mcache/mcache.go

```diff
@@ -0,0 +1,125 @@
+package mcache
+
+import (
+	"context"
+	"github.com/hashicorp/golang-lru/v2/expirable"
+	"github.com/webhookx-io/webhookx/pkg/cache"
+	"go.uber.org/zap"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	DefaultL1TTL = time.Second * 10
+	DefaultL2TTL = time.Second * 60
+)
+
+// shared instance
+var globalMcache atomic.Value
+
+func Set(mcache *MCache) {
+	globalMcache.Store(mcache)
+}
+
+// MCache is multiple levels cache
+type MCache struct {
+	mux sync.Mutex
+	l1  *expirable.LRU[string, any]
+	l2  cache.Cache
+}
+
+type Options struct {
+	L1Size int
+	L1TTL  time.Duration
+	L2     cache.Cache
+}
+
+func NewMCache(opts *Options) *MCache {
+	return &MCache{
+		l1: expirable.NewLRU[string, any](opts.L1Size, nil, opts.L1TTL),
+		l2: opts.L2,
+	}
+}
+
+func (c *MCache) InvalidateL1(ctx context.Context, key string) error {
+	c.l1.Remove(key)
+	return nil
+}
+
+func (c *MCache) InvalidateL2(ctx context.Context, key string) error {
+	return c.l2.Remove(ctx, key)
+}
+
+func (c *MCache) Invalidate(ctx context.Context, key string) error {
+	zap.S().Debugf("invalidating cache %s", key)
+	if err := c.InvalidateL2(ctx, key); err != nil {
+		return err
+	}
+	if err := c.InvalidateL1(ctx, key); err != nil {
+		return err
+	}
+	return nil
+}
+
+func Invalidate(ctx context.Context, key string) error {
+	mcache := globalMcache.Load().(*MCache)
+	return mcache.Invalidate(ctx, key)
+}
+
+type Callback[T any] func(ctx context.Context, id string) (*T, error)
+
+type LoadOptions struct {
+	DisableLRU bool
+}
+
+var defaultOpts LoadOptions
+
+func Load[T any](ctx context.Context, key string, opts *LoadOptions, cb Callback[T], id string) (*T, error) {
```

Comment:
As opts could be nil, use ` ... Option` chain to override defaultOpts for better experience.
for example:
```
func Load[T any](ctx context.Context, key string, cb Callback[T], id string, opts ...Option) (*T, error) {
```

Reply:
Adding `... Option` at the end of the function arguments make things a bit complex. I prefer to use an Options struct here.